### ISSUE TITLE
how can I download big file?--write server-sided code

### ISSUE THREAD

**Issue by iamclj** [ithewei/libhv#584]
when i finished uploading file,then ,how can I download the file--how can i write server-sided code,the follow fragment code not work correctly because of 16m limit in the function: setBody() ,that is to say,when file size less than 16m ,its ok .else it does not work.

down_my_file(HttpRequest *req, HttpResponse *resp) {
	resp->SetHeader("Content-Type", "application/octet-stream");
        std::ifstream ifs(real_uppath, std::ios::in | std::ios::binary);//real_uppath is file absolute path
try {
		char buf[8192];
		std::string rslt = "";
		while (ifs.read(buf, sizeof(buf)).gcount() > 0) {
			rslt.append(buf, ifs.gcount());
		}
		ifs.close();
		resp->SetHeader("Content-Length", std::to_string(rslt.size()));
		resp->SetBody(rslt);
 		return 200;

	} catch (std::exception &e) {
		resp->SetHeader("Content-Length", std::to_string(std::string(e.what()).size()));
		resp->SetBody(e.what());
		return 200;
	}
}

**Comment by ithewei** [comment]
You can refer to https://github.com/ithewei/libhv/blob/dd2bb958f9ef41265c66c27d3ff499993a634812/examples/httpd/handler.cpp#L276

**Comment by iamclj** [comment]
handler::sedLargefile not work correctly, the problem is still exists--when file size is small,all is ok,but when file size larger than 16m, error is the same. my client is web browser,request method is get.
   my copied code fragment:
   	router.GET("/downmyfile", [](HttpRequest *req, const HttpContextPtr &ctx) {

		int rt_code = down_my_file(req, ctx);

		return rt_code;
	});

//---------------------------------
	std::thread([ctx]() {
		ctx->writer->Begin();
		std::string filepath = real_uppath;		//real  file path
		HFile file;
		if (file.open(filepath.c_str(), "rb") != 0) {
			ctx->writer->WriteStatus(HTTP_STATUS_NOT_FOUND);
			ctx->writer->WriteHeader("Content-Type", "text/html");
			ctx->writer->WriteBody("<center><h1>404 Not Found</h1></center>");
			ctx->writer->End();
			return 200;
		}
		http_content_type content_type = CONTENT_TYPE_NONE;
		const char *suffix = hv_suffixname(filepath.c_str());
		if (suffix) {
			content_type = http_content_type_enum_by_suffix(suffix);
		}
		if (content_type == CONTENT_TYPE_NONE || content_type == CONTENT_TYPE_UNDEFINED) {
			content_type = APPLICATION_OCTET_STREAM;
		}
		size_t filesize = file.size();
		ctx->writer->WriteHeader("Content-Type", http_content_type_str(content_type));
#if USE_TRANSFER_ENCODING_CHUNKED
	        ctx->writer->WriteHeader("Transfer-Encoding", "chunked");
	#else
		ctx->writer->WriteHeader("Content-Length", filesize);
#endif
		ctx->writer->EndHeaders();

		char *buf = NULL;
		int len = 40960; // 40K
		SAFE_ALLOC(buf, len);
		size_t total_readbytes = 0;
		int last_progress = 0;
		int sleep_ms_per_send = 0;
		if (ctx->service->limit_rate <= 0) {
			// unlimited
		} else {
			sleep_ms_per_send = len * 1000 / 1024 / ctx->service->limit_rate;
		}
		if (sleep_ms_per_send == 0)
			sleep_ms_per_send = 1;
		int sleep_ms = sleep_ms_per_send;
		auto start_time = std::chrono::steady_clock::now();
		auto end_time = start_time;
		while (total_readbytes < filesize) {
			if (!ctx->writer->isConnected()) {
				break;
			}
			if (!ctx->writer->isWriteComplete()) {
				hv_delay(1);
				continue;
			}
			size_t readbytes = file.read(buf, len);
			if (readbytes <= 0) {
				// read file error!
				ctx->writer->close();
				break;
			}
			int nwrite = ctx->writer->WriteBody(buf, readbytes);
			if (nwrite < 0) {
				// disconnected!
				break;
			}
			total_readbytes += readbytes;
			int cur_progress = total_readbytes * 100 / filesize;
			if (cur_progress > last_progress) {
	 				last_progress = cur_progress;
			}
			end_time += std::chrono::milliseconds(sleep_ms);
			std::this_thread::sleep_until(end_time);
		}
		ctx->writer->End();
		SAFE_FREE(buf);
 	}).detach();
	return HTTP_STATUS_UNFINISHED;

**Comment by ithewei** [comment]
The simplest way is to use `HttpService::Static(path, dir)`. libhv implements the static resource download function internally.
